### ISSUE TITLE
Vim mode for the editor

### ISSUE THREAD

**Issue by AnderRasoVazquez** [godotengine/roadmap#9]
It would be great to have Vim style editing in the editor, because it's a really fast way to edit code. I would like to help developing it, but I don't know where to start.

**Comment by bfloch** [comment]
I second that wish.
I started to go the other direction :)
https://twitter.com/StateOffGames/status/962590588494385152

**Comment by SigmaOne** [comment]
Please, bois, support vim! Vim is life!

**Comment by Tattomoosa** [comment]
Vim mode or support would be great, but until then using VSCode with the Vim and Godot Tools extensions works pretty well.

**Comment by rosshadden** [comment]
I have to say, I have almost never used a Vim mode that came even close to replicating my workflows and productivity in Vim itself, and I would imagine most Vim users would agree with this sentiment. I honestly don't think adding a Vim mode to the official editor would make sense unless they used an existing embedded editor like [Ace](https://github.com/ajaxorg/ace), because otherwise we would never be happy with it and always want more.

I use (neo)vim for Godot, and even if they added a Vim mode that would never change. If anything I would rather they spend time improving workflows with external tools.

**Comment by TPHRyan** [comment]
As a contrasting opinion, as a person who primarily uses the JetBrains IDEA flavour of IDEs, I am completely accustomed to just using Vi mode (IDEAVim). Not because the mode is better than vim, but because it works well enough with just the basic editing bindings, buffers, searching, and macros. It's just the basic muscle-memory, deep-ingrained features that I need. Beyond that, the editor/IDE has features for more advanced stuff that is specific to the problems being solved.

Honestly, I just want to be able to use the Godot editor without needing the arrow keys. If it means I end up implementing it myself when I have more time, then so be it.

**Comment by AnderRasoVazquez** [comment]
Yes, I'm a JetBrains IDE user too. Just some basic features would be enough.

**Comment by ericmccarthy7** [comment]
Personally, I believe the best possible implementation of this is to follow the Language Server Protocol. https://microsoft.github.io/language-server-protocol/

This would create a better experience in VSCode, Atom, Vim, Emacs, and any other LSP compatible editor.

**Comment by stefanobieler** [comment]
ohh man this would be so nice. please do.

**Comment by embeddr** [comment]
This would be greatly appreciated!

The first thing I do in any IDE or editor is install a Vim hotkey plugin. If an editor doesn't have one, I typically find myself moving on to something else pretty quick! It would be excellent to have Vim hotkey support in the Godot script editor. For the time being, we'll have to make due jumping between Godot and another editor...

**Comment by Geequlim** [comment]
Why not just writing gdscript with vim or  other editors with Vim mode supported?  I don't think this kind of feature is supposed to support for a general game engine.

**Comment by pohy** [comment]
@Geequlim Because the auto completion doesn't work as nicely as in Godot. Or doesn't work at all.

**Comment by Calinou** [comment]
@pohy [A GSoC student is working on Language Server Protocol support for Godot](https://godotengine.org/article/gsoc-2019-progress-report-1-part-1#gdscript-lsp) :wink:

**Comment by netpoetica** [comment]
> I have to say, I have almost never used a Vim mode that came even close to replicating my workflows and productivity in Vim itself, and I would imagine most Vim users would agree with this sentiment.

I use Vim regularly and will say that while I agree, all "vim mode" implementations in IDEs across the board are lacking in the feature set that makes Vim amazing, the sheer amount of mouse usage I find myself doing in Godot is staggering for me. Simple Vim bindings for getting around the file without using a mouse, visual select mode for copying things, etc, are all still much better than none at all. 

I'm personally not very into exiting the Godot editor to use an external IDE, and would much prefer to stay in the zone in Godot editor but have some of the basic Vim functionality.

**Comment by Newbytee** [comment]
Has there been any code to work on implementing this?

**Comment by Jummit** [comment]
[This guy on reddit tried it.](https://www.reddit.com/r/godot/comments/3lha11/how_to_integrate_vim_into_godot_engine/)

**Comment by woshichuanqilz** [comment]
Supported now? Seems so long time passed...

**Comment by Jummit** [comment]
> Supported now? Seems so long time passed...

It's currently not supported.
